### PR TITLE
Merge purchaseFor / purchaseForFrom

### PR DESCRIPTION
**Pull request by NickCuso**
# Description

Merging purchaseFor and purchaseForFrom into a single `purchase` function.

# Issues

Fixes #
Refs #2182

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [ ] My code follows the style guidelines of this project, including naming conventions
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 1]
```diff
@@ -22,53 +22,22 @@ contract MixinPurchase is
...
-    hasValidKey(_referrer)
-  {
-    return _purchaseFor(_recipient, _referrer);
-  }
-
-  /**
-  * @dev Purchase function: this lets a user purchase a key from the lock for another user
-  * @param _recipient address of the recipient of the purchased key
   * This will fail if
   *  - the keyReleaseMechanism is private
```
Comment (nfurfaro) [inline smart-contracts/contracts/mixins/MixinPurchase.sol]: We could probably remove references to the "keyReleaseMechanism" as well( if we're not going to be introducing private or restricted locks). These are old and confusing to someone who doesn't know what they are.

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 2]
```diff
@@ -22,53 +22,22 @@ contract MixinPurchase is
...
   *  - the keyReleaseMechanism is Approved and the recipient has not been previously approved
   *  - the amount value is smaller than the price
   *  - the recipient already owns a key
-  * TODO: next version of solidity will allow for message to be added to require.
   */
-  function _purchaseFor(
+  function purchase(
     address _recipient,
     address _referrer
```
Comment (nfurfaro) [inline smart-contracts/contracts/mixins/MixinPurchase.sol]: So fo r a purchase with no referrer, a 0 has to be passed in as the second param here right? I'm on board with pushing that responsibility to the front end in order to simplify things in the contract.

**Comment by nfurfaro** [comment]
Just s thought here, but could we not delete all of the tests with names starting with "Non_Public..."?

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 1 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: Hmm yea.  Actually none of those points are valid.  Deleted.

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 2 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: That's correct

**Comment by NickCuso** [comment]
We could start to rename the files, but we should be careful about removing them unless we are sure the scenario is covered in a different test.

**Pull request merged**